### PR TITLE
[pytorch] In ProcessGroupAgent, use non-iostream torch::load()/save().

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28063 [pytorch] In ProcessGroupAgent, use non-iostream torch::load()/save().**

Avoid using the iostream versions of torch::load()/torch::save(), which
incur at least one additional full data copy.

Differential Revision: [D17945206](https://our.internmc.facebook.com/intern/diff/D17945206/)

### CODE REVIEW

**Comment by mrshenli** [comment]
Thanks for adding this optimization! 

Test failures are real:

```
Oct 16 08:14:41 Running test_rpc_fork ... [2019-10-16 08:14:41.900323]
Oct 16 08:21:22 test_add (__main__.RpcTestWithFork) ... ERROR
Oct 16 08:28:03 test_add_with_id (__main__.RpcTestWithFork) ... ERROR
Oct 16 08:28:03 test_asymmetric_load_with_join (__main__.RpcTestWithFork)
Oct 16 08:34:43 Test graceful termination. ... ERROR
Oct 16 08:41:23 test_async_add (__main__.RpcTestWithFork) ... ERROR
Oct 16 08:48:04 test_builtin_remote_ret (__main__.RpcTestWithFork) ... ERROR
....
```

**Comment by jjlilley** [comment]
Thanks, Shen.

I see the caffe2/test:rpc_fork timeouts, looking at them locally now.

Part of the motivation is that I have a parallel thrift-based diff (with some test coverage, but definitely worse), and I want to make sure there are no gotchas with the new codepath.

**Comment by jjlilley** [comment]
fwiw, I think this should be ready for review - I had a type mismatch on the templated Storage data().

I recently landed a separately parallel change for ThriftRpcAgent (D17944157); it's mostly only going to affect big payloads, but in the case of Thrift, the benchmarks were 20-30% range.

**Review by mrshenli — APPROVED** [review]
LGTM!

File: torch/csrc/distributed/rpc/process_group_agent.cpp [hunk 1]
```diff
@@ -27,13 +27,27 @@ void serialize(const Message& message, std::ostream& os) {
   // append id and autograd metadata as a tensor
   tensors.push_back(torch::tensor({message.id()}, {torch::kInt64}));
 
-  torch::save(tensors, os);
+  // optional: estimate output size, to avoid some unnecessary resizing.
+  static constexpr size_t kBaseOverhead = 2048;
+  static constexpr size_t kPerTensor = 128;
```
[the remark this answers is not in the document]
Comment (jjlilley) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: Thanks for looking!

tensors[i].nbytes() is effectively numel * element_size() , i.e. from TensorBody.h:
  size_t nbytes() const {                                                                                                                                        
      return impl_->numel() * impl_->itemsize();                                                                                                                   
    }       

The two magic numbers are mostly representative of typical fixed + per tensor overhead imposed by the current zip format used by torch::save().  (kind of higher than many folks might expect)

They slightly overestimate, but not by much (e.g. kBaseOverhead looks closer to 1600 bytes in practice), and in practice, underestimating is definitely noticeable in the benchmarks.

[the remark this answers is not in the document]
Comment (mrshenli) [reply torch/csrc/distributed/rpc/process_group_agent.cpp]: ah, I see, I missed the `nbytes()` call.

**Pull request closed without merging**